src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java

```diff
@@ -386,9 +386,20 @@ public ClusterState execute(final ClusterState currentState) throws Exception {
                                 if (mergeResult.hasConflicts()) {
                                     throw new MergeMappingException(mergeResult.buildConflicts());
                                 }
+                            } else {
+                                // TODO: can we find a better place for this validation?
```

Comment:
I tried to move to the MapperService#merge(...) but that didn't work out for create index api. In the case an new index is created with a parent and child mapping defined in the same request, it can fail if the parent type is defined first in the mappings part of the request body, it fails then with the validation error we are adding here. This validation shouldn't be checked in case if mappings are defined in a create index api call, because it is a new index.

This validation should only be checked if the put mapping api is used, but in `MapperService#merge(...)` we can't tell if we got invoked from put mapping or create index api. The place where the validation is now, we can make sure the validation is only checked if the put mapping api is invoked. I'm still not liking the place for this validation... but without a refactoring I don't see how we can move this check to the mapper service